File: python/exercise/polymorph.py
class Animal:
    def __init__(self, name):
        self.name = name


class Cat(Animal):
    def __init__(self, name):
        super().__init__(name)

    def meow(self):
        print(f'{self.name} - meow')


class Dog(Animal):
    def __init__(self, name):
        super().__init__(name)

    def woof(self):
        print(f'{self.name} - woof')


class Human:
    def __init__(self, name):
        self.name = name

    def do_something(self):
        print(f'{self.name} do something')


class Student(Human):
    def __init__(self, name):
        super().__init__(name)

    def do_homework(self):
        print(f'{self.name} do homework')


class Teacher(Human):
    def __init__(self, name):
        super().__init__(name)

    def teaching(self):
        print(f'{self.name} teaching')


def print_human_names(l: list[Human]) -> None:
    for i in l:
        print(i.name)


def print_animal_names(l: list[Animal]) -> None:
    for i in l:
        print(i.name)


if __name__ == '__main__':
    l_human: list[Human] = [Student('Dima'), Teacher('Tatiana')]
    l_animal: list[Animal] = [Dog('Sharik'), Cat('Murzik')]

    print(l_animal)
    print(l_human)

    print_human_names(l_human)
    print('-----------')
    print_animal_names(l_animal)

    print('===========')

    print_human_names(l_animal)
    print('-----------')
    print_animal_names(l_human)

File: scala/something-small/src/main/scala/org/dedkot/Polymorph.scala
package org.dedkot

case class Animal(name: String)

class Dog(name: String) extends Animal(name) {
  def woof(): Unit = print(s"$name - woof")
}

class Cat(name: String) extends Animal(name) {
  def meow(): Unit = print(s"$name - meow")
}

case class Human(val name: String)

class Student(name: String) extends Human(name) {
  def doHomework(): Unit = print(s"$name do homework")
}

class Teacher(name: String) extends Human(name) {
  def teaching(): Unit = print(s"$name teaching")
}

object Polymorph extends App {
  def printNames[A](l: List[A]): Unit = l.foreach {
    case Animal(name) => println(name)
    case Human(name)  => println(name)
    case _            => "WTF?!"
  }

  val lHumans: List[Human]   = List(new Student("Dima"), new Teacher("Tatiana"))
  val lAnimals: List[Animal] = List(new Dog("Sharik"), new Cat("Murzik"))

  println(lHumans)
  println(lAnimals)

  printNames(lHumans)
  println("------------")
  printNames(lAnimals)

  println("============")

  printNames(lAnimals)
  println("------------")
  printNames(lHumans)
}
